File: backend/src/main/java/com/mycompany/online_shop_backend/controllers/handler/ControllerAdvice.java
package com.mycompany.online_shop_backend.controllers.handler;

import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import javax.servlet.http.HttpServletRequest;

@Slf4j
@RestControllerAdvice
public class ControllerAdvice {

    @ExceptionHandler
    public ResponseEntity<ErrorResponse> unauthorizedException(UsernameNotFoundException ex, HttpServletRequest request) {
        var errorResponse = new ErrorResponse(
                HttpStatus.UNAUTHORIZED, ex.getMessage(), request.getRequestURI()
        );
        return new ResponseEntity<>(errorResponse, HttpStatus.UNAUTHORIZED);
    }

    @ExceptionHandler
    public ResponseEntity<ErrorResponse> exception(Exception ex, HttpServletRequest request) {
        HttpStatus httpStatus = HttpStatus.INTERNAL_SERVER_ERROR;
        ResponseStatus responseStatus = ex.getClass().getAnnotation(ResponseStatus.class);
        if (responseStatus != null) {
            httpStatus = responseStatus.value();
        }

        switch (httpStatus) {
            case INTERNAL_SERVER_ERROR -> log.error("", ex);
            case BAD_REQUEST -> log.debug("400 Bad Request - " + ex);
            case UNAUTHORIZED -> log.debug("Unauthorized");
            default -> log.warn("", ex);
        }

        var errorResponse = new ErrorResponse(
                httpStatus, ex.getMessage(), request.getRequestURI()
        );
        return new ResponseEntity<>(errorResponse, httpStatus);
    }
}

File: book-service/src/main/kotlin/com/example/bookservice/api/rest/error/ControllerAdvice.kt
package com.example.bookservice.api.rest.error

import com.example.bookservice.api.rest.model.ErrorCode
import com.example.bookservice.api.rest.model.ErrorResponse
import io.github.oshai.kotlinlogging.KotlinLogging
import jakarta.servlet.http.HttpServletRequest
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import java.time.OffsetDateTime
import java.util.*

@RestControllerAdvice
class ControllerAdvice {

    companion object {
        private val logger = KotlinLogging.logger(this::class.java.name)
    }

    @ExceptionHandler
    fun serviceException(ex: ServiceException, request: HttpServletRequest): ResponseEntity<ErrorResponse> =
        ResponseEntity(
            ErrorResponse(
                errorId = UUID.randomUUID(),
                timestamp = OffsetDateTime.now(),
                path = request.requestURI,
                code = ex.errorCode,
                message = ex.message!!
            ),
            ex.httpStatusCode
        )

    @ExceptionHandler
    fun exception(ex: Exception, request: HttpServletRequest): ResponseEntity<ErrorResponse> {
        val response = ResponseEntity(
            ErrorResponse(
                errorId = UUID.randomUUID(),
                timestamp = OffsetDateTime.now(),
                path = request.requestURI,
                code = ErrorCode.INTERNAL_SERVER_ERROR,
                // message must be logged, but not included in response due to security reasons
                message = "Internal Server Error"
            ),
            HttpStatus.INTERNAL_SERVER_ERROR
        )
        logger.error { "Error: ${ex.message}, response: $response" }
        return response
    }
}